### PR TITLE
Python 3: Convert some unicode/bytes uses

### PR DESCRIPTION
**Pull request by hawkowl**
Also fixes https://github.com/matrix-org/synapse/issues/3306 by actually normalising passwords. This has the potential for breaking people's passwords, but no more than the issue of using a different browser or device breaking someone's password.

### CODE REVIEW

File: synapse/storage/events.py [hunk 1]
```diff
@@ -65,7 +65,10 @@
 
 
 def encode_json(json_object):
-    return frozendict_json_encoder.encode(json_object)
+    if PY3:
+        return frozendict_json_encoder.encode(json_object)
+    else:
+        return frozendict_json_encoder.encode(json_object).decode("utf-8")
```
[the remark this answers is not in the document]
Comment (hawkowl) [reply synapse/storage/events.py]: json/simplejson returns `str` (so unicode on py3, bytes on py2). We want it as unicode, so we have to decode it on Py2.

File: synapse/handlers/auth.py [hunk 2]
```diff
@@ -855,8 +856,16 @@ def hash(self, password):
             Deferred(str): Hashed password.
         """
         def _do_hash():
-            return bcrypt.hashpw(password.encode('utf8') + self.hs.config.password_pepper,
-                                 bcrypt.gensalt(self.bcrypt_rounds))
+            # Ensure that we normalise the password
+            if isinstance(password, bytes):
```
[the remark this answers is not in the document]
Comment (hawkowl) [reply synapse/handlers/auth.py]: That would be a nicer way of writing it, but password isn't in a writable scope (because it's inside another function).

[the remark this answers is not in the document]
Comment (hawkowl) [reply synapse/handlers/auth.py]: wrt the type -- it's being passed in as str, but since it's not clear that it would only be decoded on py2, I've changed it to be a Py3 check.

File: synapse/http/server.py [hunk 3]
```diff
@@ -296,8 +298,14 @@ def _async_render(self, request):
         # here. If it throws an exception, that is handled by the wrapper
         # installed by @request_handler.
 
+        def _parse(s):
+            if PY3:
```
[the remark this answers is not in the document]
Comment (hawkowl) [reply synapse/http/server.py]: Given the URL encoded sequence string "%E2%98%83", encoded as a unicode string:

Python2:
```
>>> urllib.unquote(u"%E2%98%83")
u'\xe2\x98\x83'
```
This is wrong, it returns Unicode but the ASCII escaped character codes.

Python 2:
```
urllib.unquote(u"%E2%98%83".encode('ascii')).decode('utf8')
u'\u2603'
```
Correct, returns the Unicode literal, but escaped for display as Py2 will usually not print real Unicode characters by itself.

Python 3:
```
>>> urllib.parse.unquote(u"%E2%98%83")
'☃'
```
Correct, returns the Unicode literal (not escaped, as Python 3 has the correct terminal encoding set up).

File: synapse/http/server.py [hunk 4]
```diff
@@ -327,7 +335,7 @@ def _get_handler_for_request(self, request):
         # Loop through all the registered callbacks to check if the method
         # and path regex match
         for path_entry in self.path_regexs.get(request.method, []):
-            m = path_entry.pattern.match(request.path)
+            m = path_entry.pattern.match(request.path.decode())
```
[the remark this answers is not in the document]
Comment (hawkowl) [reply synapse/http/server.py]: fixed

File: synapse/http/server.py [hunk 5]
```diff
@@ -383,7 +391,7 @@ def __init__(self, path):
         self.url = path
 
     def render_GET(self, request):
-        return redirectTo(self.url, request)
+        return redirectTo(self.url.encode(), request)
```
[the remark this answers is not in the document]
Comment (hawkowl) [reply synapse/http/server.py]: fixed

File: synapse/storage/signatures.py [hunk 6]
```diff
@@ -82,7 +82,16 @@ def _get_event_reference_hashes_txn(self, txn, event_id):
             " WHERE event_id = ?"
         )
         txn.execute(query, (event_id, ))
-        return {k: v for k, v in txn}
+        if six.PY2:
+            return {k: v for k, v in txn}
+        else:
+            done = {}
+            for k, v in txn:
+                if not isinstance(v, bytes):
```
[the remark this answers is not in the document]
Comment (hawkowl) [reply synapse/storage/signatures.py]: It shouldn't, so I removed the if.

File: synapse/storage/signatures.py [hunk 7]
```diff
@@ -82,7 +82,16 @@ def _get_event_reference_hashes_txn(self, txn, event_id):
...
+                    done[k] = v.encode('ascii')
```
[the remark this answers is not in the document]
Comment (hawkowl) [reply synapse/storage/signatures.py]: done

File: synapse/http/server.py [hunk 8]
```diff
@@ -296,8 +298,14 @@ def _async_render(self, request):
         # here. If it throws an exception, that is handled by the wrapper
         # installed by @request_handler.
 
+        def _parse(s):
+            if PY3:
+                return urllib.parse.unquote(s)
+            else:
+                return urllib.parse.unquote(s.encode('utf8')).decode('utf8')
```
[the remark this answers is not in the document]
Comment (hawkowl) [reply synapse/http/server.py]: We aren't encoding this because we now parse the incoming arguments to Unicode in `_get_handler_for_request`.

[the remark this answers is not in the document]
Comment (hawkowl) [reply synapse/http/server.py]: (as to why we encode it on py2, see https://github.com/matrix-org/synapse/pull/3569/files/4831ead29a575a299d8e377f26bc8fc35dc49cc2#r205017485 )

File: synapse/storage/events.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (hawkowl) [reply synapse/storage/events.py]: (also the decodes in the functions below have been moved to this one spot)

File: synapse/handlers/auth.py [hunk 9]
```diff
@@ -626,6 +629,10 @@ def validate_login(self, username, login_submission):
         # special case to check for "password" for the check_password interface
         # for the auth providers
         password = login_submission.get("password")
+
+        if password and PY2:
+            password = password.decode('utf8')
```
Comment (richvdh) [inline synapse/handlers/auth.py]: why is this necessary? surely login_submission is coming from the json body so should already have been un-utf8'ed?

(and if it is necessary, why don't we do it on PY3?)

Comment (richvdh) [reply synapse/handlers/auth.py]: (should it not just be `password = unicode(password)` ?)

File: synapse/handlers/auth.py [hunk 10]
```diff
@@ -708,6 +715,7 @@ def _check_local_password(self, user_id, password):
 
         Args:
             user_id (str): complete @user:id
+            password (unicode): the provided password
```
Comment (richvdh) [inline synapse/handlers/auth.py]: I think it might be worth clarifying that it will be a `str` on python3. Something like:

```
password (str|unicode): the provided password. On python2, *must* be a unicode.
```

(and similar on `validate_hash` etc etc)

Of course what's really happening is that elsewhere we have been sloppy about saying `str` when we mean `str|unicode`, though I don't suggest we change that...

File: synapse/handlers/register.py [hunk 11]
```diff
@@ -131,7 +131,7 @@ def register(
         Args:
             localpart : The local part of the user ID to register. If None,
               one will be generated.
-            password (str) : The password to assign to this user so they can
+            password (unicode) : The password to assign to this user so they can
```
Comment (richvdh) [inline synapse/handlers/register.py]: again, `(str|unicode)` with words of explanation

File: synapse/http/server.py [hunk 12]
```diff
@@ -296,8 +298,14 @@ def _async_render(self, request):
         # here. If it throws an exception, that is handled by the wrapper
         # installed by @request_handler.
 
+        def _unquote(s):
+            if PY3:
+                return urllib.parse.unquote(s)
```
Comment (richvdh) [inline synapse/http/server.py]: why do we not utf-8 decode here?

File: synapse/http/server.py [hunk 13]
```diff
@@ -296,8 +298,14 @@ def _async_render(self, request):
...
+            else:
+                return urllib.parse.unquote(s.encode('ascii')).decode('utf8')
```
Comment (richvdh) [inline synapse/http/server.py]: why do we `encode('ascii')` here?

File: synapse/rest/client/v1/admin.py [hunk 14]
```diff
@@ -175,6 +175,7 @@ def on_POST(self, request):
         from synapse.rest.client.v2_alpha.register import RegisterRestServlet
         register = RegisterRestServlet(self.hs)
 
+        password = password.decode('utf-8')
```
Comment (richvdh) [inline synapse/rest/client/v1/admin.py]: if there's a reason for doing `password.decode('utf-8')` rather than `body["password"]`, can you add a comment to say what it is?

File: synapse/rest/client/v2_alpha/account.py [hunk 15]
```diff
@@ -165,6 +166,9 @@ def on_POST(self, request):
         assert_params_in_dict(params, ["new_password"])
         new_password = params['new_password']
 
+        if PY2:
+            new_password = new_password.decode('utf8')
```
Comment (richvdh) [inline synapse/rest/client/v2_alpha/account.py]: again, why not `unicode(new_password)` ?

File: synapse/rest/client/v2_alpha/register.py [hunk 16]
```diff
@@ -389,8 +389,13 @@ def on_POST(self, request):
             assert_params_in_dict(params, ["password"])
 
             desired_username = params.get("username", None)
-            new_password = params.get("password", None)
             guest_access_token = params.get("guest_access_token", None)
+            new_password = params.get("password", None)
+
+            if new_password and isinstance(new_password, bytes):
+                # We may not need to decode the password, if it came from the
```
Comment (richvdh) [inline synapse/rest/client/v2_alpha/register.py]: this smells wrong. Surely we should be storing the same type in the session as is in the original params?

Comment (richvdh) [reply synapse/rest/client/v2_alpha/register.py]: also, haven't `params` already been decoded?

File: synapse/storage/events.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (richvdh) [reply synapse/storage/events.py]: ok comments please to explain this.

File: synapse/handlers/auth.py (hunk 9 above)
Comment (hawkowl) [reply synapse/handlers/auth.py]: `unicode(password)` also will decode it with ASCII, as it's forced decoding.

File: synapse/handlers/auth.py (hunk 10 above)
Comment (hawkowl) [reply synapse/handlers/auth.py]: It's clearer to use `bytes` (flat bytes), `str` (str on either platform), or `unicode` (unicode), even though there is no such thing as `unicode` on Python 3.

File: synapse/http/server.py (hunk 12 above)
Comment (hawkowl) [reply synapse/http/server.py]: It's already decoded by `_get_handler_for_request`.

File: synapse/http/server.py (hunk 13 above)
Comment (hawkowl) [reply synapse/http/server.py]: URLs are 7-bit ASCII.

File: synapse/rest/client/v2_alpha/register.py (hunk 16 above)
Comment (hawkowl) [reply synapse/rest/client/v2_alpha/register.py]: The original params are `str` (so bytes on Python 2) because json.loads returns strs, not Unicode, we don't want that.

File: synapse/handlers/auth.py (hunk 10 above)
Comment (richvdh) [reply synapse/handlers/auth.py]: ok, as discussed, `unicode` seems the right answer here. However, the difference with `user_id` and the return type is now very striking; suggest you update them too.

File: synapse/storage/signatures.py [hunk 17]
```diff
@@ -74,15 +74,21 @@ def _get_event_reference_hashes_txn(self, txn, event_id):
             txn (cursor):
             event_id (str): Id for the Event.
         Returns:
-            A dict of algorithm -> hash.
+            A dict[str, bytes] of algorithm -> hash.
```
Comment (richvdh) [inline synapse/storage/signatures.py]: s/str/unicode/ ?

File: synapse/storage/signatures.py [hunk 18]
```diff
@@ -74,15 +74,21 @@ def _get_event_reference_hashes_txn(self, txn, event_id):
...
         """
         query = (
             "SELECT algorithm, hash"
             " FROM event_reference_hashes"
             " WHERE event_id = ?"
         )
         txn.execute(query, (event_id, ))
-        return {k: v for k, v in txn}
+        if six.PY2:
```
Comment (richvdh) [inline synapse/storage/signatures.py]: can we not use the PY3 path under py2 as well?

File: synapse/storage/signatures.py [hunk 19]
```diff
@@ -74,15 +74,21 @@ def _get_event_reference_hashes_txn(self, txn, event_id):
...
+            return {k: v for k, v in txn}
+        else:
+            done = {}
+            for k, v in txn:
+                done[k] = v.encode('ascii')
```
Comment (richvdh) [inline synapse/storage/signatures.py]: fwiw I'd find a dict comprehension clearer here:

```
return {k: v.encode('ascii') for k, v in txn}
```

File: synapse/util/frozenutils.py [hunk 20]
```diff
@@ -66,5 +66,7 @@ def _handle_frozendict(obj):
 
 # A JSONEncoder which is capable of encoding frozendics without barfing
 frozendict_json_encoder = json.JSONEncoder(
+    ensure_ascii=False,
+    encoding="utf8",
```
Comment (richvdh) [inline synapse/util/frozenutils.py]: this is the default \<shrug>

File: synapse/util/frozenutils.py [hunk 21]
```diff
@@ -66,5 +66,7 @@ def _handle_frozendict(obj):
 
 # A JSONEncoder which is capable of encoding frozendics without barfing
 frozendict_json_encoder = json.JSONEncoder(
+    ensure_ascii=False,
```
Comment (richvdh) [inline synapse/util/frozenutils.py]: I don't think this should be necessary (and it makes simplejson much slower). Could you add comments to explain why it is necessary?

File: synapse/http/server.py (hunk 13 above)
Comment (richvdh) [reply synapse/http/server.py]: ugh, what a mess. thanks for clarifying this.

File: synapse/util/frozenutils.py (hunk 21 above)
Comment (richvdh) [reply synapse/util/frozenutils.py]: as discussed, let's just coerce the result to unicode
